code/FileAttachmentFieldTrack.php

```diff
@@ -0,0 +1,69 @@
+<?php
+
+/**
+ * Track files as they're uploaded and remove when they've been saved.
+ *
+ * @package  unclecheese/silverstripe-dropzone
+ */
+class FileAttachmentFieldTrack extends DataObject {
+    private static $db = array(
+        'SessionID' => 'Varchar(255)',
+        'ControllerClass' => 'Varchar(60)',
+        'RecordID' => 'Int',
+        'RecordClass' => 'Varchar(60)',
+    );
+
+    private static $has_one = array(
+        'File' => 'File',
+    );
+
+    public static function untrack($fileIDs) {
+        if (!$fileIDs) {
+            return;
+        }
+        $fileIDs = (array)$fileIDs;
+        $trackRecords = FileAttachmentFieldTrack::get()->filter(array('FileID' => $fileIDs));
+        foreach ($trackRecords as $trackRecord) {
```

Comment:
I dug into DataList on 3.2. I'm not making that change. If you follow 'removeAll' codepath you'll see it iterates over each record, and ends up re-querying each record to delete it.

That's just stupid.